File: Robotics Kinematics and Dynamics Course/Chapter8/Lesson5/Chapter8_Lesson5.m
function cost = avoidanceCost2R(q, l1, l2, qmin, qmax, eps, eps_ws)
% q: [q1; q2]
if nargin < 2, l1 = 1.0; end
if nargin < 3, l2 = 1.0; end
if nargin < 4, qmin = [-pi; -pi]; end
if nargin < 5, qmax = [ pi;  pi]; end
if nargin < 6, eps  = 1e-3; end
if nargin < 7, eps_ws = 1e-3; end

q1 = q(1); q2 = q(2);

% Jacobian
s1  = sin(q1);  c1  = cos(q1);
s12 = sin(q1+q2); c12 = cos(q1+q2);
J = [ -l1*s1 - l2*s12, -l2*s12;
       l1*c1 + l2*c12,  l2*c12 ];

% Manipulability
JJt = J*J.';
w = sqrt(det(JJt));
phi_sing = 1/(w^2 + eps);

% Joint-limit penalty
phi_joint = 0;
for i = 1:2
    qi = q(i); ql = qmin(i); qu = qmax(i);
    phi_joint = phi_joint + 1/(qi - ql)^2 + 1/(qu - qi)^2;
end

% Workspace penalty
x = l1*cos(q1) + l2*cos(q1+q2);
y = l1*sin(q1) + l2*sin(q1+q2);
r = hypot(x, y);
rmin = abs(l1 - l2);
rmax = l1 + l2;
d_ws = min(r - rmin, rmax - r);
phi_ws = 1/(d_ws^2 + eps_ws);

cost = phi_sing + phi_joint + phi_ws;
end
      

File: Tutorials/Codes/Robotics_Kinematics_and_Dynamics_Course/Chapter8/Lesson5/Chapter8_Lesson5.py
import numpy as np

# 2R Jacobian
def jacobian_2r(q, l1=1.0, l2=1.0):
    q1, q2 = q
    s1, c1 = np.sin(q1), np.cos(q1)
    s12, c12 = np.sin(q1 + q2), np.cos(q1 + q2)

    J = np.array([
        [-l1 * s1 - l2 * s12, -l2 * s12],
        [ l1 * c1 + l2 * c12,  l2 * c12]
    ], dtype=float)
    return J

def manipulability(J):
    JJt = J @ J.T
    return float(np.sqrt(np.linalg.det(JJt)))

def joint_limit_penalty(q, q_min, q_max):
    q = np.asarray(q, dtype=float)
    q_min = np.asarray(q_min, dtype=float)
    q_max = np.asarray(q_max, dtype=float)
    phi = 0.0
    for qi, ql, qu in zip(q, q_min, q_max):
        phi += 1.0 / (qi - ql)**2 + 1.0 / (qu - qi)**2
    return phi

def fk_2r(q, l1=1.0, l2=1.0):
    q1, q2 = q
    x = l1 * np.cos(q1) + l2 * np.cos(q1 + q2)
    y = l1 * np.sin(q1) + l2 * np.sin(q1 + q2)
    return np.array([x, y], dtype=float)

def workspace_penalty(q, l1=1.0, l2=1.0, eps_ws=1e-3):
    x = fk_2r(q, l1, l2)
    r = np.linalg.norm(x)
    r_min = abs(l1 - l2)
    r_max = l1 + l2
    d_ws = min(r - r_min, r_max - r)
    return 1.0 / (d_ws**2 + eps_ws)

def avoidance_cost(q,
                   l1=1.0, l2=1.0,
                   q_min=(-np.pi, -np.pi),
                   q_max=( np.pi,  np.pi),
                   eps=1e-3):
    J = jacobian_2r(q, l1, l2)
    w = manipulability(J)
    phi_sing = 1.0 / (w**2 + eps)
    phi_joint = joint_limit_penalty(q, q_min, q_max)
    phi_ws = workspace_penalty(q, l1, l2)
    return phi_sing + phi_joint + phi_ws

def finite_diff_grad(f, q, h=1e-4):
    q = np.asarray(q, dtype=float)
    g = np.zeros_like(q)
    for i in range(len(q)):
        dq = np.zeros_like(q)
        dq[i] = h
        g[i] = (f(q + dq) - f(q - dq)) / (2.0 * h)
    return g

def gradient_step(q, step=1e-2):
    # One offline gradient step to reduce avoidance cost
    cost_fun = lambda x: avoidance_cost(x)
    g = finite_diff_grad(cost_fun, np.array(q, dtype=float))
    return np.array(q, dtype=float) - step * g

if __name__ == "__main__":
    q = np.array([0.0, 0.5])
    for k in range(20):
        c = avoidance_cost(q)
        print(f"Iter {k}: q = {q}, cost = {c:.6f}")
        q = gradient_step(q, step=5e-2)
      
